pkg/commands/command_helper.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2020-Present VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package commands
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/pivotal/build-service-cli/pkg/k8s"
+)
+
+type CommandHelper struct {
+	dryRun bool
+	output bool
+	wait   bool
+
+	outWriter io.Writer
+	errWriter io.Writer
+
+	objPrinter k8s.ObjectPrinter
+	strBuilder strings.Builder
+}
+
+func NewCommandHelper(cmd *cobra.Command) (*CommandHelper, error) {
+	dryRun, err := getBoolFlag("dry-run", cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := getStringFlag("output", cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	wait, err := getBoolFlag("wait", cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	var objPrinter k8s.ObjectPrinter
+
+	outputResource := len(output) > 0
+	if outputResource {
+		objPrinter, err = k8s.NewObjectPrinter(output)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &CommandHelper{
+		dryRun:     dryRun,
+		output:     outputResource,
+		wait:       wait,
+		outWriter:  cmd.OutOrStdout(),
+		errWriter:  cmd.ErrOrStderr(),
+		objPrinter: objPrinter,
+		strBuilder: strings.Builder{},
+	}, nil
+}
+
+func (ch CommandHelper) IsDryRun() bool {
+	return ch.dryRun
+}
+
+func (ch CommandHelper) ShouldWait() bool {
+	return ch.wait && !ch.dryRun && !ch.output
+}
+
+func (ch CommandHelper) PrintObjs(objs []runtime.Object) error {
+	if ch.output {
```

Comment:
It might be interesting to have a nullObjectPrinter that does not print anything to avoid this `if`

Reply:
Started out with using a similar concept as you suggested (DiscardWriter), but took it out to limit further delegation and keep the CommandHelper simple. And even with a DiscardWriter, it still goes through the motions but nulls it out at the end. The flag on the other hand is an earlier termination. In some ways, not using further delegation allows for the flag based behavioural change to be documented within CommandHelper

If you still see value in using a nullObjectPrinter / DiscardWriter, happy to make the change.

Reply:
I was just thinking out loud. 🙂

Reply:
aye. the NullHandler was a very tempting pattern, indeed. Stepped back from it to limit hops

Reply:
marking as resolved for now. happy to reopen if required.